### ISSUE TITLE
Modules emitted in the wrong order (Library file not found: why3.Testbin_Impl0_Transition_Interface)

### ISSUE THREAD

**Issue by gmorenz** [xldenis/creusot#552]
The following code ends up emitting `Testbin_Impl0_Transition_Interface` before it's used in `Testbin_Impl1_Step`. This results in the error message `Library file not found: why3.Testbin_Impl0_Transition_Interface` when trying to open it in why3. Simply cut-and-pasting the Interface to before it's used in the file fixes the problem.

```rust
use creusot_contracts::{logic, ensures};

pub struct Transition;
pub struct Machine;

#[logic]
pub fn transition_logical(this: &Machine) -> bool {
    this.description_logical();
    true
}

impl Machine {
    #[ensures(transition_logical(self))]
    pub fn transition(&self) -> Transition {
        Transition
    }
}

pub trait MachineTrait {
    #[logic]
    fn description_logical(self);

    fn step(&mut self);
}

impl MachineTrait for Machine {
    #[logic]
    fn description_logical(self) {}

    fn step(&mut self) {
        self.transition();
    }
}

fn main() {}
```

[And the .mlcfg (with .log added to the filename because github filters file uploads by extension](https://github.com/xldenis/creusot/files/9537655/testbin-bin.mlcfg.log)

Related to #457.

**Comment by xldenis** [comment]
There must be a missing dependency edge. I really want to overhaul the handling of all of this soon..

**Comment by gmorenz** [comment]
Better reproduction, this one not just emits things in the wrong order, it also fails to emit a `with` clause which makes it un-provable - which means I can't just work around this with a script that re-arranges modules.

```rust
use creusot_contracts::{predicate, maintains, requires, ensures};

pub struct Transition;
pub struct Machine;

pub trait MachineTrait {
    #[predicate]
    fn invariants(self) -> bool;

    #[maintains((mut self).invariants())]
    fn step(&mut self) -> bool;
}

impl MachineTrait for Machine {
    #[predicate]
    fn invariants(self) -> bool {
        true
    }

    #[maintains((mut self).invariants())]
    fn step(&mut self) -> bool {
        self.transition(); // Comment out this line and it works
        false
    }
}

impl Machine {
    #[requires(self.invariants())]
    pub fn transition(&self) -> Transition {
        Transition
    }
}
```

My current understand of the problem is that:

1. We start translating a function, and even generates dependencies for it, [here](https://github.com/xldenis/creusot/blob/92abb6ce2f0c06c277bb66b055daed70b809e2f0/creusot/src/ctx.rs#L136)

2. Before storing the dependencies for the function, we recurses down and starts emitting the functions used [here](https://github.com/xldenis/creusot/blob/92abb6ce2f0c06c277bb66b055daed70b809e2f0/creusot/src/ctx.rs#L158). 

3. That second function tries to find the dependencies for the first function, and fails, [here](https://github.com/xldenis/creusot/blob/adf83838953dbb20d34c6a1f011011c3e9e6994c/creusot/src/clone_map.rs#L399). It treats this as if it has no dependencies.

4. We finally store the dependencies we should have found in the above step, [here](https://github.com/xldenis/creusot/blob/92abb6ce2f0c06c277bb66b055daed70b809e2f0/creusot/src/ctx.rs#L167).

In the happy path we don't recurse to something that uses our dependencies, so step 4 occurs before step 3 and everything works.

I think a short sighted solution to the problem is to just store the dependencies separately from the translated items, so we can immediately store them upon generating them, instead of only after we've finished recursing. I don't think that should a very invasive change, so I'll take a stab at implementing it shortly. It's not clear to me what forces these translate calls to be called in the correct order though, it wouldn't surprise me if this is just kicking the bug down the line.

I suppose if this code is going to be rewritten soon, that might be good enough.

**Comment by xldenis** [comment]
Ah I see, there's a cycle induced between the trait implementation and the function `transition`. I don't think Creusot can properly handle this today because we need there to exist an acyclic ordering of all the functions and clones. 

However, Creusot _should_ properly detect this error case and report it clearly. That should be easy enough to add. 
When I overhaul cloning I may also be able to enable mutually recursive functions as a side-effect.

**Comment by xldenis** [comment]
> I think a short sighted solution to the problem is to just store the dependencies separately from the translated items

I think that is a good idea and I'm planning on doing this anyways. I'd like the core datastructure of creusot to keep track of all translated items as a graph rather than an `IndexMap`. This would also enable generating multiple whyml files per crate (for example one per function).

**Comment by gmorenz** [comment]
I'm not sure I agree there's an actual cycle. I think it depends on whether you view calling `machine.invariants()` as depending on the impl block or just the specific method, and creusot seems to view it as the latter. Which should result in the following acyclic dependency graph:

![graphviz](https://user-images.githubusercontent.com/595972/189735963-614f8d98-2fca-4781-a124-e9b543a8344e.svg)

As mentioned in the PR (#554), fixing the dependency issue (step depends on invariants) doesn't fix the ordering issue. But it also doesn't appear to create a cyclic dependency graph (i.e. if I uncomment [the assert checking for cycles](https://github.com/xldenis/creusot/blob/adf83838953dbb20d34c6a1f011011c3e9e6994c/creusot/src/clone_map.rs#L588) it doesn't find any in the above code). So I think there might be another issue here...

Don't feel like you need to prioritize this, I can wait (and/or continue trying to fix things myself), deadlines don't :)

**Comment by xldenis** [comment]
Hmm, well then perhaps splitting out dependencies is all we need to do!

**Comment by xldenis** [comment]
I just hit this same bug 😬.

**Comment by xldenis** [comment]
Ok, so I understand _why_ there's an issue here.

During the translation of `transition`, we use `self.invariant`, which during cloning triggers a check to see if there are any laws we should clone. This check asks for the impl to be translated, which causes the ordering issue we see here. If the trait had laws you could see a cycle in the dependencies, but since there aren't any, you don't. 

I see the solution being two things:
- Keep track of the stack of things being translated, since we recurse during the translation of one item to launch the translation of another, we should track that stack and error when we see a cycle in it (aka `step` being translated twice simultaneously).
- Use a graph to order the translated modules. This will make us less sensitive to the exact order each `translate` call is issued in and would seamlessly handle this case.
